esp-hal/src/aes/mod.rs

```diff
@@ -595,6 +751,480 @@ pub mod dma {
+    ///         b'd', 0,
+    ///     ],
+    /// );
+    ///
+    /// // Process a block of data in this context. The ECB mode of operation requires that
+    /// // the length of the data is a multiple of the block (16 bytes) size.
+    /// let input_buffer: [u8; 16] = [
+    ///     b'm', b'e', b's', b's', b'a', b'g', b'e', 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    /// ];
+    /// let mut output_buffer: [u8; 16] = [0; 16];
+    ///
+    /// let operation_handle = ecb_encrypt
+    ///     .process(&input_buffer, &mut output_buffer)
+    ///     .unwrap();
+    /// operation_handle.wait_blocking();
+    ///
+    /// // output_buffer now contains the ciphertext
+    /// assert_eq!(
+    ///     output_buffer,
+    ///     [
+    ///         0xb3, 0xc8, 0xd2, 0x3b, 0xa7, 0x36, 0x5f, 0x18, 0x61, 0x70, 0x0, 0x3e, 0xd9, 0x3a,
+    ///         0x31, 0x96,
+    ///     ]
+    /// );
+    /// #
+    /// # {after_snippet}
+    /// ```
+    pub struct AesDmaBackend<'d> {
+        peri: AES<'d>,
+        dma: PeripheralDmaChannel<AES<'d>>,
+        driver: Option<AesDma<'d>>,
+        state: DmaState,
+    }
+
+    impl<'d> AesDmaBackend<'d> {
+        #[procmacros::doc_replace(
+            "dma_channel" => {
+                cfg(esp32s2) => "let dma_channel = peripherals.DMA_CRYPTO;",
+                _ => "let dma_channel = peripherals.DMA_CH0;"
+            }
+        )]
+        /// Creates a new DMA-enabled AES backend.
+        ///
+        /// The backend needs to be [`start`][Self::start]ed before it can execute AES operations.
+        ///
+        /// ## Example
+        ///
+        /// ```rust, no_run
+        /// # {before_snippet}
+        /// use esp_hal::aes::dma::AesDmaBackend;
+        ///
+        /// # {dma_channel}
+        /// let mut aes = AesDmaBackend::new(peripherals.AES, dma_channel);
+        /// # {after_snippet}
+        /// ```
+        pub fn new(aes: AES<'d>, dma: impl DmaChannelFor<AES<'d>>) -> Self {
+            Self {
+                peri: aes,
+                dma: dma.degrade(),
+                driver: None,
+                state: DmaState::Idle,
+            }
+        }
+
+        #[procmacros::doc_replace(
+            "dma_channel" => {
+                cfg(esp32s2) => "let dma_channel = peripherals.DMA_CRYPTO;",
+                _ => "let dma_channel = peripherals.DMA_CH0;"
+            }
+        )]
+        /// Registers the DMA-driven AES driver to process AES operations.
+        ///
+        /// The driver stops operating when the returned object is dropped.
+        ///
+        /// ## Example
+        ///
+        /// ```rust, no_run
+        /// # {before_snippet}
+        /// use esp_hal::aes::dma::AesDmaBackend;
+        ///
+        /// # {dma_channel}
+        /// let mut aes = AesDmaBackend::new(peripherals.AES, dma_channel);
+        /// let _handle = aes.start();
+        /// # {after_snippet}
+        /// ```
+        pub fn start(&mut self) -> AesDmaWorkQueueDriver<'_, 'd> {
```

Comment:
Another tangential question. Should there be a `stop()` or is `drop()` the way?

Reply:
I need to figure out actually suspending the driver, and I guess that will need blocking & async stops.